_drafts/measuring-stack-usage.md

```diff
@@ -0,0 +1,329 @@
+functionality, to give a reasonably accurate measurement of the maximum stack
+usage.
+
+_Note: since this approach relies on values being written to the stack, if a
+buffer is allocated on stack but only partially used, the full buffer size
+wouldn't be immediately represented in the fill pattern._
+
+Another application of this technique is to use it measure stack consumption of
+a particular function call, with the following procedure:
+
+1. Just before the function call, re-apply the stack fill pattern across the
+   unused portion of the stack
+2. Save the current stack pointer value, and call the function of interest
+3. Measure the stack high watermark, and compute the difference between the
+   saved stack pointer value and the high watermark. This is the stack usage of
+   the function
+
+### Runtime analysis with GDB single-stepping
+
+This is a technique I applied recently to measure stack usage without modifying
+the target device code.
+
+The idea is to use GDB to single-step through the program, and to dump the stack
+pointer register, `$sp`, at every step. I used a small GDB Python script to do
+this work, a small snippet is shown below:
+
+```python
+# save the starting stack pointer value
+sp_start = int(gdb.parse_and_eval("$sp"))
+print("Starting SP: 0x{:08x}".format(sp_start))
+
+# record the lowest (memory index) value- this is the max stack usage for the
+# function. initialize it to the maximum addressable address.
+sp_min = 0xFFFFFFFF
+while True:
+    # single instruction step, so we capture _any_ sp change
+    gdb.execute("si")
+    sp = int(gdb.parse_and_eval("$sp"))
+    if sp < sp_min:
+        sp_min = sp
+    # the full implementation exits this loop when hitting a breakpoint
+    if hit_breakpoint():
+      break
+print("Min SP: 0x{:08x}".format(sp_min))
+print("Delta SP: {}".format(sp_start - sp_min))
+```
+
+The script is _very slow_ to run since it interrupts program flow at every
+instruction, but it gives a very accurate result.
+
+It can be enhanced to produce more data, for example recording the max usage for
+each frame in a call stack:
+
+```python
+# write the verbose data to a file
+with open("stack_usage.log", "w") as f:
+    sp_min = 0xFFFFFFFF
+    current_function = None
+    current_function_max_su = 0
+    while True:
+        gdb.execute("si")
+        sp = int(gdb.parse_and_eval("$sp"))
+        su = sp_start - sp
+        name = gdb.selected_frame().name()
+        # if we just changed function name, record the high water mark
+        # from the last function
+        if current_function is not None and name != current_function:
+            f.write(f"{current_function}|{current_function_max_su}\n")
+            current_function_max_su = 0
+        current_function = name
+        if su > current_function_max_su:
+            current_function_max_su = su
+        # f.write(f"{current_function}|{current_function_max_su}\n")
```

Comment:
Should this be uncommented?

Reply:
ah that was for debug, good spot! deleting it